crud/common/sharding_key.lua

```diff
@@ -0,0 +1,122 @@
+        if fieldno == nil then
+            return nil, GetShardingKeyFieldnosError:new(
+                "No such field (%s) in a space format (%s)", field_name, space_name)
+        end
+        table.insert(fieldnos, {fieldno = fieldno})
+    end
+
+    return {parts = fieldnos}
+end
+
+-- It's a public method to access it in tests.
+function sharding_key_module.fetch_on_storage()
+    local sharding_key_def = {}
+    local sharding_key_space = box.space._ddl_sharding_key
+    if sharding_key_space == nil then
+        return sharding_key_def
+    end
+
+    local SPACE_NAME_FIELDNO = 1
+    local SPACE_SHARDING_KEY_FIELDNO = 2
+    for _, tuple in ipairs(sharding_key_space:select()) do
+        local space_name = tuple[SPACE_NAME_FIELDNO]
+        local key = tuple[SPACE_SHARDING_KEY_FIELDNO]
+        sharding_key_def[space_name] = key
+    end
+
+    return sharding_key_def
+end
+
+local sharding_key_fetch_in_progress = false
+local sharding_key_fetch_cond = fiber.cond()
+
+-- Under high load we may get a case when many fibers will fetch sharding keys
+-- from storages. It is not good from performance point of view.
+-- concurrent_fetch() wraps a fetch_on_storage() to limit a number of fibers
+-- that fetches a sharding key by a single one, other fibers will wait for a
+-- FETCH_SHARDING_KEY_TIMEOUT and then make one more attempt to fetch sharding
+-- key if a cache is still empty.
+local function concurrent_fetch(replicaset)
+    dev_checks('table')
+
+    if sharding_key_fetch_in_progress == true and
+       not sharding_key_fetch_cond:wait(const.FETCH_SHARDING_KEY_TIMEOUT) then
+        return nil, FetchShardingKeyError:new(
+                        'Waiting for sharding key to be fetched is timed out')
+    end
+
+    local sharding_key_def = {}
+    if sharding_key_fetch_in_progress == false then
+        sharding_key_fetch_in_progress = true
+        local err
+        sharding_key_def, err = replicaset:call(FETCH_FUNC_NAME, {},
+                                                {timeout = const.FETCH_SHARDING_KEY_TIMEOUT})
+        if err ~= nil then
+            return nil, err
+        end
+        sharding_key_fetch_cond:broadcast()
+        sharding_key_fetch_in_progress = false
+    end
+
+    return sharding_key_def
+end
+
+-- Get sharding key definitions (field names) for all spaces or a certain space
+-- (when space_name is not nil). Function uses a cache that organized as a table in
+-- following format:
+-- {
+--     'space_name_1' = {'name', 'age'},
+--     'space_name_2' = {'age'},
+--     ...
+-- }
+-- Return an sharding key or nil when space_name is specified or a whole map
+-- with all sharding keys.
+function sharding_key_module.fetch_on_router(space_name)
+    dev_checks('string')
+
+    if cache.sharding_key_def == nil then
+        local replicasets = vshard.router.routeall()
+        local replicaset = select(2, next(replicasets))
+        local err
+        cache.sharding_key_def, err = concurrent_fetch(replicaset)
```

Comment:
From private dialogs with @ligurio. Pseudocode:

```lua
if has cache then
    return from cache
end
while fetch_in_progress do
    wait
    if has cache then
        return from cache
    end
    if timeout then
        return nil, err
    end
end

fetch_in_progress = true
fetch()
fetch_in_progress = false
broadcast()
if timeout then
    return nil, err
end
fill cache
return from cache
```

It is a bit weird that we'll mix locking, caching and fetching within one function. However it looks inevitable. If we'll move caching upward, we'll perform unnecessary fetches (like it is implemented now). If we'll move fetching upward, we'll lock around cache accesses and not fetching itself — that's senseless. If we'll move locking upward, we'll lock cache accesses that is costly.

Reply:
```lua
while fetch_in_progress do
    wait
    if has cache then
        return from cache
    end
    if timeout then
        return nil, err
    end
end
```
Loop is useless here because we will exit from it as soon as timeout is exceeded. We can add a counter and make for example three attempts with timeout, however I would make one loop so I replaced `while do` with `if then`.

Reply:
Rewrote `reentrant_fetch()` body again.

Changes:
- added loop
- corrected naming (`timeout` → `timeout_is_exceeded`)
- removed condition for `metadata_map == nil` because `fetch_on_storage()` always returns a table
- marked critical section with comments for better readability
- changed timeouts, now total sum of all timeouts could not bigger than `FETCH_SHARDING_KEY_TIMEOUT`
- moved filling cache to the critical section (to avoid inconsistency due to `yield()`)

Reply:
Patch:

```diff
diff --git a/crud/common/sharding_key.lua b/crud/common/sharding_key.lua
index bfeaa8b..44887b8 100644
--- a/crud/common/sharding_key.lua
+++ b/crud/common/sharding_key.lua
@@ -71,33 +71,35 @@ local function reentrant_fetch()
         return cache.sharding_key_as_index_obj_map
     end
 
-    if sharding_key_fetch_in_progress == true then
-        local timeout = false
-        if sharding_key_fetch_cond:wait(const.FETCH_SHARDING_KEY_TIMEOUT) then
-            timeout = true
+    local timeout_deadline = fiber.clock() + const.FETCH_SHARDING_KEY_TIMEOUT
+    local timeout_is_exceeded
+    while sharding_key_fetch_in_progress == true do
+        timeout_is_exceeded = false
+        local timeout = timeout_deadline - fiber.clock()
+        if sharding_key_fetch_cond:wait(timeout) then
+            timeout_is_exceeded = true
         end
         if cache.sharding_key_as_index_obj_map ~= nil then
             return cache.sharding_key_as_index_obj_map
         end
-        if timeout then
+        if timeout_is_exceeded == true then
             return nil, FetchShardingKeyError:new(
-                            'Waiting for sharding key to be fetched is timed out')
+                            'Waiting for sharding key to be fetched is timed out (%d sec)',
+                            const.FETCH_SHARDING_KEY_TIMEOUT)
         end
     end
 
-    sharding_key_fetch_in_progress = true
+    sharding_key_fetch_in_progress = true -- Start of critical section.
     local metadata_map, err = call.any(FETCH_FUNC_NAME, {}, {
-        timeout = const.FETCH_SHARDING_KEY_TIMEOUT
+        timeout = timeout_deadline - fiber.clock()
     })
     if err ~= nil then
-        sharding_key_fetch_in_progress = false
+        sharding_key_fetch_in_progress = false -- End of critical section.
         sharding_key_fetch_cond:broadcast()
         return nil, err
     end
 
-    sharding_key_fetch_in_progress = false
-    sharding_key_fetch_cond:broadcast()
-
+    -- metadata_map is nil when fetch_on_storage() is timed out.
     if metadata_map == nil then
         return {}
     end
@@ -112,6 +114,9 @@ local function reentrant_fetch()
         sharding_key_as_index_obj_map[space_name] = sharding_key_as_index_obj
     end
 
+    sharding_key_fetch_in_progress = false -- End of critical section.
+    sharding_key_fetch_cond:broadcast()
+
     return sharding_key_as_index_obj_map
 end
 ```

Reply:
> * moved filling cache to the critical section (to avoid inconsistency due to `yield()`)

It is not updated ATM on the branch? I see that we fill the cache in `sharding_key_module.fetch_on_router()` currently.

Another thing that I would highlight is that we should not exceed given timeout for the whole request (such as `crud.insert()`): so if we're going to fetch sharding key, we should perform next storage request with lower timeout. We can solve it separately, I don't think that it is critical.

Reply:
It seems, we can protect our critical section similarly to the `locked()` pattern that can be found in src/box/lua/load_cfg.lua in tarantool. It uses 1-sized channel and looks really tiny. And we'll not need to spin in the loop. Several points around:

* It worth to store the channel within the sharding_key.cache module, together with the cache itself. This way it should work good with hot reloading: behaviour will not depend on order of reloading.
* We should check cache presence before stepping into the critical section and right after it: a kind of double checked locking pattern.

Please, take care to timeout situation: how we should catch it and what next fibers will do after timeout fail in current one. The same for 'no sharding keys info' situation.

Reply:
> It seems, we can protect our critical section similarly to the locked() pattern that can be found in src/box/lua/load_cfg.lua in tarantool. It uses 1-sized channel and looks really tiny. And we'll not need to spin in the loop.  Several points around:
> - It worth to store the channel within the sharding_key.cache module, together with the cache itself. This way it should work good with hot reloading: behaviour will not depend on order of reloading.
> - We should check cache presence before stepping into the critical section and right after it: a kind of double checked locking pattern.
> Please, take care to timeout situation: how we should catch it and what next fibers will do after timeout fail in current one. The same for 'no sharding keys info' situation.

I did my best to implement locking based on fiber channel as you proposed.

```diff
+++ b/crud/common/sharding_key.lua
@@ -15,6 +15,30 @@ local FETCH_FUNC_NAME = '_crud.fetch_on_storage'
 
 local sharding_key_module = {}
 
+-- Function decorator that is used to prevent _fetch_on_router() from being
+-- called concurrently by different fibers.
+local function locked(f, timeout)
+    dev_checks('function', 'number')
+    return function(...)
+        local timeout_deadline = fiber.clock() + timeout
+        local ok = cache.fetch_lock:put(true, timeout)
+        -- channel:put() returns false in two cases: when timeout is exceeded
+        -- or channel has been closed. However error message describes only
+        -- first reason, I'm not sure we need to disclose to users such details
+        -- like problems with synchronization objects.
+        if not ok then
+            return FetchShardingKeyError:new(
+                "Timeout for fetching sharding key for space '%s' is exceeded", space_name)
+        end
+        local timeout = timeout_deadline - fiber.clock()
+        local status, err = pcall(f, timeout, ...)
+        cache.fetch_lock:get()
+        if not status or err ~= nil then
+            return err
+        end
+    end
+end
+
```

Full patch with changes is not top of the branch.